### PR TITLE
Added a test for QuerySet.select_related() reverse caching.

### PR DESCRIPTION
**Pull request by ticosax**
Taken from #9081

### CODE REVIEW

File: tests/queries/tests.py [hunk 1]
```diff
@@ -1717,6 +1717,16 @@ def test_tickets_3045_3288(self):
...
 
+    def test_select_related_reverse_caching(self):
+        t1 = Tag.objects.create(name='t1')
+        t2 = Tag.objects.create(name='t2', parent=t1)
+        tag = Tag.objects.select_related('parent').filter(name=t2.name).first()
+        with self.assertNumQueries(0):
+            self.assertEqual(tag.parent.name, t1.name)
+        self.assertNotIn('children', tag.parent._state.fields_cache)
+        with self.assertNumQueries(1):
+            self.assertEqual(list(tag.parent.children.all())[0].name, t2.name)
```
Comment (charettes) [inline tests/queries/tests.py]: You could use `tag.parent.first().name` here instead.

**Comment by timgraham** [comment]
As far as I can tell, the test passes on master without the code changes in this PR.

**Comment by ticosax** [comment]
This refactoring is needed to support `FilteredRelation`.
You ask me to submit as a separate PR. I'm confused.

**Comment by timgraham** [comment]
The test passes without the code changes so it's unclear to me what its purpose is and why it should be in the same commit as the refactorings.

**Comment by ticosax** [comment]
I don't know the reason, maybe @holvianssi had to write this test and iterate several times until he found the right approach to refactor `select_related`. Probably the coverage has been increased after that, even if no regression has been detected by the refactoring. This refactoring is not to fix a bug but to extend its internal API so `FilteredRelation` can be supported.
Maybe you'd prefer not having that test ?

**Comment by timgraham** [comment]
The idea would be to add the test only in a commit (if we can confirm it's adding some additional coverage). The test is probably be better organized in the `select_related`, `select_related_onetoone`, or `select_related_regress` test app.

**Comment by holvianssi** [comment]
The test is about calling the setters unconditionally in populate() (note the removed if condition at location of calling remote_setter()). Django shouldn't set anything in the reverse direction for multivalued relations. At least an earlier iteration of the filtered relation patch did that. So, in essence it's a regression test for a bug that doesn't exist in master branch.

**Comment by timgraham** [comment]
Thanks, Anssi, I found the reason for the test. In the related PR, there's a line `'remote_setter': f.remote_field.set_cached_value if f.unique else lambda x, y: None`. If the if branch is removed there. this test is the only test in Django's suite that will fail.  I adjusted the test so it's in the `select_related` test app.

**Pull request merged**